core/src/miner/mem_pool.rs

```diff
@@ -269,15 +269,103 @@ impl MemPool {
+        //FIXME: change this random to be reproducible
+        let body_length = rand::thread_rng().gen_range(50, 200);
+        let random_bytes = (0..body_length).map(|_| rand::random::<u8>()).collect();
+        Transaction::new("Sample".to_string(), random_bytes)
+    }
+
+    #[test]
+    fn remove_all() {
+        let validator = Arc::new(TestCoordinator::default());
+        let db = Arc::new(kvdb_memorydb::create(crate::db::NUM_COLUMNS.unwrap_or(0)));
+        let mut mem_pool = MemPool::with_limits(8192, usize::max_value(), db, validator);
+
+        let inserted_block_number = 1;
+        let inserted_timestamp = 100;
+        let origin = TxOrigin::External;
+
+        let transactions: Vec<_> = (0..10).map(|_| create_random_transaction()).collect();
+
+        let add_result = mem_pool.add(transactions.clone(), origin, inserted_block_number, inserted_timestamp);
+        assert!(add_result.iter().all(|r| r.is_ok()));
+
+        mem_pool.remove_all();
+        assert!(transactions.iter().all(|tx| { !mem_pool.transaction_pool.contains(&tx.hash()) }));
+        assert_eq!(mem_pool.transaction_pool.len(), 0);
+        assert_eq!(mem_pool.transaction_pool.count, 0);
+        assert_eq!(mem_pool.transaction_pool.mem_usage, 0);
+    }
+
     #[test]
     fn add_and_remove_transactions() {
-        //TODO: Write test after implementing a mockup coordinator
-        todo!()
+        let validator = Arc::new(TestCoordinator::default());
+        let db = Arc::new(kvdb_memorydb::create(crate::db::NUM_COLUMNS.unwrap_or(0)));
+        let mut mem_pool = MemPool::with_limits(8192, usize::max_value(), db, validator);
+
+        let inserted_block_number = 1;
+        let inserted_timestamp = 100;
+        let origin = TxOrigin::External;
+
+        let transactions: Vec<_> = (0..10).map(|_| create_random_transaction()).collect();
+
+        let add_result = mem_pool.add(transactions.clone(), origin, inserted_block_number, inserted_timestamp);
+        assert!(add_result.iter().all(|r| r.is_ok()));
+
+        let (to_remove, to_keep) = transactions.split_at(5);
+
+        let to_remove_hashes: Vec<_> = to_remove.iter().map(|tx| tx.hash()).collect();
+        let to_keep_hashes: Vec<_> = to_keep.iter().map(|tx| tx.hash()).collect();
+        mem_pool.remove(&to_remove_hashes);
+
+        assert!(to_keep_hashes.iter().all(|hash| { mem_pool.transaction_pool.contains(hash) }));
+        assert!(to_remove_hashes.iter().all(|hash| { !mem_pool.transaction_pool.contains(hash) }));
+
+        let count: usize = 5;
+        let mem_usage: usize = to_keep.iter().map(|tx| tx.size()).sum();
+
+        assert_eq!(mem_pool.transaction_pool.count, count);
+        assert_eq!(mem_pool.transaction_pool.mem_usage, mem_usage);
     }
 
     #[test]
     fn db_backup_and_recover() {
-        //TODO: Write test after implementing a mockup coordinato
-        todo!()
+        let validator = Arc::new(TestCoordinator::default());
```

Comment:
Why don't you implement it in 92cb444287c161f2c4d4fa010236cc5ed107eeba?

Reply:
I commented it because I think the test is not compiled in 92cb444287c161f2c4d4fa010236cc5ed107eeba. If it can be compiled, please ignore this comment.

Reply:
The answer is similar to the previous comment. Implementing `TestCoordinator` before this commit would be quite costly.

I didn't consider whether the test can be compiled or not, because even the main code was not compiling in the commit.
The commit would become too big if I had tried to make each commit compilable.